sql/bulkmerge: TestWaitForViablePlan failed

sql/bulkmerge.TestWaitForViablePlan [failed](https://mesolite.cluster.engflow.com/invocations/default/d74a7071-7fe2-4204-a15d-2e5b6814b1ea?testReportRun=10&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2J1bGttZXJnZTpidWxrbWVyZ2VfdGVzdA==) on master @ [8cf7d10d20eac900a3cda4306090d7b02ae716e7](https://github.com/cockroachdb/cockroach/commits/8cf7d10d20eac900a3cda4306090d7b02ae716e7):


```
=== RUN   TestWaitForViablePlan
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestWaitForViablePlan1934491164
    test_log_scope.go:82: use -show-logs to present logs inline
    merge_test.go:1103: -- test log scope end --
test logs left over in: outputs.zip/logTestWaitForViablePlan1934491164
--- FAIL: TestWaitForViablePlan (0.22s)
=== RUN   TestWaitForViablePlan/retries_until_coverage_is_reached
    merge_test.go:1065: 
        	Error Trace:	pkg/sql/bulkmerge/merge_test.go:1065
        	Error:      	Received unexpected error:
        	            	distributed merge failed: required SQL instances remain uncovered after 100ms: distributed merge requires SST files from instances [2] which are unavailable
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/bulkmerge.waitForViablePlan
        	            	  | 	pkg/sql/bulkmerge/merge.go:280
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/bulkmerge.TestWaitForViablePlan.func2
        	            	  | 	pkg/sql/bulkmerge/merge_test.go:1063
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:2036
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1771
        	            	Wraps: (2) distributed merge failed: required SQL instances remain uncovered after 100ms
        	            	Wraps: (3) forced error mark
        	            	  | "permanent job error"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (4) distributed merge requires SST files from instances [2] which are unavailable
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *markers.withMark (4) *bulkmerge.InstanceUnavailableError
        	Test:       	TestWaitForViablePlan/retries_until_coverage_is_reached
--- FAIL: TestWaitForViablePlan/retries_until_coverage_is_reached (0.11s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWaitForViablePlan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-64235

--- a/pkg/sql/bulkmerge/merge_test.go
+++ b/pkg/sql/bulkmerge/merge_test.go
@@ -1026,11 +1026,20 @@ func TestWaitForViablePlan(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	tightRetry := retry.Options{
+	// riggedRetry is used by subtests that intentionally exhaust the retry
+	// budget so the test itself doesn't block for long.
+	riggedRetry := retry.Options{
 		InitialBackoff: 1 * time.Millisecond,
 		MaxBackoff:     5 * time.Millisecond,
 		Multiplier:     2,
 		MaxDuration:    100 * time.Millisecond,
+	}
+	// generousRetry is used by subtests that expect the retry loop to succeed.
+	generousRetry := retry.Options{
+		InitialBackoff: 1 * time.Millisecond,
+		MaxBackoff:     5 * time.Millisecond,
+		Multiplier:     2,
+		MaxDuration:    30 * time.Second,
 	}
 
 	sstsRequiringInstance2 := []execinfrapb.BulkMergeSpec_SST{
@@ -1045,7 +1054,7 @@ func TestWaitForViablePlan(t *testing.T) {
 		}
 
 		_, _, err := waitForViablePlan(
-			context.Background(), sstsRequiringInstance2, tightRetry, plan)
+			context.Background(), sstsRequiringInstance2, generousRetry, plan)
 		require.NoError(t, err)
 		require.Equal(t, 1, attempts)
 	})
@@ -1061,7 +1070,7 @@ func TestWaitForViablePlan(t *testing.T) {
 		}
 
 		_, _, err := waitForViablePlan(
-			context.Background(), sstsRequiringInstance2, tightRetry, plan)
+			context.Background(), sstsRequiringInstance2, generousRetry, plan)
 		require.NoError(t, err)
 		require.Equal(t, 2, attempts)
 	})
@@ -1072,7 +1081,7 @@ func TestWaitForViablePlan(t *testing.T) {
 		}
 
 		_, _, err := waitForViablePlan(
-			context.Background(), sstsRequiringInstance2, tightRetry, plan)
+			context.Background(), sstsRequiringInstance2, riggedRetry, plan)
 		require.Error(t, err)
 		var unavailErr *InstanceUnavailableError
 		require.True(t, errors.As(err, &unavailErr))
@@ -1086,7 +1095,7 @@ func TestWaitForViablePlan(t *testing.T) {
 		}
 
 		_, _, err := waitForViablePlan(
-			context.Background(), sstsRequiringInstance2, tightRetry, plan)
+			context.Background(), sstsRequiringInstance2, generousRetry, plan)
 		require.ErrorIs(t, err, boom)
 	})
 
@@ -1097,7 +1106,7 @@ func TestWaitForViablePlan(t *testing.T) {
 			return nil, []base.SQLInstanceID{1, 3}, nil // never covers 2
 		}
 
-		_, _, err := waitForViablePlan(ctx, sstsRequiringInstance2, tightRetry, plan)
+		_, _, err := waitForViablePlan(ctx, sstsRequiringInstance2, generousRetry, plan)
 		require.ErrorIs(t, err, context.Canceled)
 	})
 }

